### PR TITLE
[vulkan] Release timepoint fences more aggressively when possible

### PR DESCRIPTION
**Pull request by antiagainst**
In the emulated timeline semaphore, we use VkFence as a way to
probe the status of submissions. This is needed for both the
emulated timeline semaphores and the serializing queue itself
because we can wait on either the timeline semaphore or the queue.
This requires us to hold ref ptrs to the timepoint fence in both
the timeline semaphore and the queue. The expectation is that when
both releases the ref ptr, the fence can be automatically released
to the pool.

However, the ref ptr held in the queue can typically linger on
longer because we normally don't wait queue/device idle until
the very last when all work are done. Because of the ref hold
there, the fences will not be released.

This commit adds another callback to let the emulated timeline
semaphore to inform the queue when some fences are known to
have signaled (because the timeline signaled passing a certain
value tracked by the fence), so that we can more aggressively
drop the ref ptr inside the serializing queue.

At the same time, reduces the number of initial VkFence requested.

Fixes https://github.com/google/iree/issues/2526

### CODE REVIEW

**Comment by antiagainst** [comment]
Note that this should fix #2526; the failure is actually another issue tracked by #2659..

File: iree/hal/vulkan/emulated_timeline_semaphore.cc [hunk 1]
```diff
@@ -200,24 +204,42 @@ StatusOr<VkSemaphore> EmulatedTimelineSemaphore::GetSignalSemaphore(
 
 StatusOr<bool> EmulatedTimelineSemaphore::TryToAdvanceTimeline(
     uint64_t to_upper_value) {
-  IREE_TRACE_SCOPE0("EmulatedTimelineSemaphore::TryToAdvanceTimeline");
+  absl::InlinedVector<VkFence, 4> signaled_fences;
+  auto status = TryToAdvanceTimeline(to_upper_value, &signaled_fences);
```
Comment (ScottTodd) [inline iree/hal/vulkan/emulated_timeline_semaphore.cc:208]: Why defer checking the status until after calling `on_fence_signal_`?

**Comment by antiagainst** [comment]
@hanhanW FYI

File: iree/hal/vulkan/emulated_timeline_semaphore.cc (hunk 1 above)
Comment (antiagainst) [reply iree/hal/vulkan/emulated_timeline_semaphore.cc:208]: I don't think it matters that much? If errors happen, either returning immediately or still sending the queue a notice to clean up something are both okay to me..

File: integrations/tensorflow/e2e/BUILD [hunk 2]
```diff
@@ -80,7 +80,6 @@ LLVM_FAILING = [
 # keep sorted
 VULKAN_FAILING = [
     "broadcasting_test.py",
-    "control_flow_test.py",
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply integrations/tensorflow/e2e/BUILD:83]: Actually I was wrong.This reveals a different bug than #2659. I've #2905 to fix that.

**Comment by antiagainst** [comment]
Thanks @ScottTodd! :D

**Comment by ScottTodd** [comment]
> Thanks @ScottTodd! :D

No problem, sorry for letting this sit for so long x_x

**Comment by antiagainst** [comment]
No worries. :) Gonna be cautious on changes related to timeline semaphore. ;-P

**Pull request merged**